File: backend/app/schemas/social.py
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel


class ConnectionStatus(str, Enum):
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    ERROR = "error"


class ConnectionResponse(BaseModel):
    id: str
    platform: str
    status: ConnectionStatus
    handle: Optional[str] = None
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None
    followers: Optional[int] = None
    scopes: List[str] = []
    needs_reauth: bool = False
    connected_at: Optional[datetime] = None
    last_synced_at: Optional[datetime] = None
    last_error: Optional[str] = None
    workspace_id: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ConnectionList(BaseModel):
    items: List[ConnectionResponse]
    connected: List[str]
    provider: str = "upload_post"
    configured: bool = True


class ConnectRequest(BaseModel):
    platform: str


class ConnectStartResponse(BaseModel):
    authorize_url: str
    state: Optional[str] = None
    expires_in: Optional[int] = None


class HostedConnectResponse(BaseModel):
    url: str

File: frontend/lib/api/social.ts
import { http } from "./http";

export type ConnectionStatus = "connected" | "disconnected" | "error";

/** Platforms Tolkyn can connect + publish to. */
export const SUPPORTED_PLATFORMS = [
  "facebook",
  "instagram",
  "x",
  "linkedin",
  "tiktok",
  "youtube",
] as const;

export interface SocialConnection {
  id: string;
  platform: string;
  status: ConnectionStatus;
  handle: string | null;
  display_name: string | null;
  avatar_url: string | null;
  followers: number | null;
  scopes: string[];
  needs_reauth: boolean;
  connected_at: string | null;
  last_synced_at: string | null;
  last_error: string | null;
  workspace_id: string;
  created_at: string;
  updated_at: string;
}

export interface ConnectionList {
  items: SocialConnection[];
  connected: string[];
  provider: string;
  configured: boolean;
}

export interface ConnectStart {
  authorize_url: string;
  state: string | null;
  expires_in: number | null;
}

export const socialApi = {
  list: () => http.get<ConnectionList>("/accounts"),
  sync: () => http.post<ConnectionList>("/accounts/sync"),
  /** Begin the OAuth flow for one platform — returns the URL to send the browser to. */
  connect: (platform: string) => http.post<ConnectStart>("/accounts/connect", { platform }),
  /** Branded hosted page where the user can connect / remove any channel. */
  connectPage: () => http.get<{ url: string }>("/accounts/connect-page"),
  /** Marks the platform disconnected locally and returns the hosted page URL to finish revoking. */
  disconnect: (platform: string) => http.del<{ url: string }>(`/accounts/${platform}`),
};
